src/modules/cdp/cdp_tls.c

```diff
@@ -0,0 +1,196 @@
+	tls_methods[TLS_USE_TLSv1_2 - 1].TLSMethodMax = TLS1_2_VERSION;
+
+	/* ranges of TLS versions (require a minimum TLS version) */
+	tls_methods[TLS_USE_TLSv1_PLUS - 1].TLSMethod = TLS_method();
+	tls_methods[TLS_USE_TLSv1_PLUS - 1].TLSMethodMin = TLS1_VERSION;
+
+	tls_methods[TLS_USE_TLSv1_1_PLUS - 1].TLSMethod = TLS_method();
+	tls_methods[TLS_USE_TLSv1_1_PLUS - 1].TLSMethodMin = TLS1_1_VERSION;
+
+	tls_methods[TLS_USE_TLSv1_2_PLUS - 1].TLSMethod = TLS_method();
+	tls_methods[TLS_USE_TLSv1_2_PLUS - 1].TLSMethodMin = TLS1_2_VERSION;
+}
+
+/*
+ * Convert TLS method string to integer
+ */
+int tls_parse_method(str* m)
+{
+	cfg_option_t* opt;
+
+	if (!m) {
+		LM_BUG("Invalid parameter value\n");
+		return -1;
+	}
+
+	opt = cfg_lookup_token(methods, m);
+	if (!opt) return -1;
+
+	return opt->val;
+}
+
+SSL_CTX* init_ssl_ctx(int method)
+{
+	SSL_CTX* ctx;
+
+	SSL_library_init();
+	OpenSSL_add_all_algorithms();
+	SSL_load_error_strings();
+
+	/* libssl >= 1.1.0 */
+	ctx = SSL_CTX_new(tls_methods[method - 1].TLSMethod);
+	if (ctx == NULL) {
+		unsigned long e = 0;
+		e = ERR_peek_last_error();
+		LM_ERR("init_ssl_ctx(): Failed to create SSL context (%lu: %s / %s)\n",
+				e, ERR_error_string(e, NULL),
+				ERR_reason_error_string(e));
+		return NULL;
+	}
+
+	if(method>TLS_USE_TLSvRANGE) {
+		if(tls_methods[method - 1].TLSMethodMin) {
+			SSL_CTX_set_min_proto_version(ctx,
+					tls_methods[method - 1].TLSMethodMin);
+		}
+	} else {
+		if(tls_methods[method - 1].TLSMethodMin) {
+			SSL_CTX_set_min_proto_version(ctx,
+					tls_methods[method - 1].TLSMethodMin);
+		}
+		if(tls_methods[method - 1].TLSMethodMax) {
+			SSL_CTX_set_max_proto_version(ctx,
+					tls_methods[method - 1].TLSMethodMax);
+		}
+	}
+	return ctx;
+}
+
+int load_certificates(SSL_CTX* ctx, str* cert, str* key)
+{
+	str cert_fixed = STR_NULL;
+	str key_fixed = STR_NULL;
+
+	if (pkg_str_dup(&cert_fixed, cert) < 0) {
+		LM_ERR("load_certificates(): Failed to copy cert parameter\n");
+		return -1;
+	}
+	if (pkg_str_dup(&key_fixed, key) < 0) {
+		LM_ERR("load_certificates(): Failed to copy key parameter\n");
+		return -1;
+	}
+	if (SSL_CTX_use_certificate_file(ctx, cert_fixed.s, SSL_FILETYPE_PEM) <= 0) {
+		LM_ERR("load_certificates(): Unable to load certificate file\n");
+		return -1;
+	}
+	if (SSL_CTX_use_PrivateKey_file(ctx, key_fixed.s, SSL_FILETYPE_PEM) <= 0) {
+		LM_ERR("load_certificates(): Unable to load private key file\n");
+		return -1;
+	}
+	if (!SSL_CTX_check_private_key(ctx)) {
+		LM_ERR("load_certificates(): Private key does not match the public certificate\n");
+		return -1;
+	}
+	return 0;
+}
+
+SSL* init_ssl_conn(int client_fd, SSL_CTX *ctx)
+{
+	/* Create an SSL object */
+	SSL *ssl = SSL_new(ctx);
+	if (!ssl) {
+		LM_ERR("init_ssl_conn(): Failed to create SSL object.");
+		SSL_CTX_free(ctx);
+		return NULL;
+	}
+	/* Set up the TLS connection */
+	if (SSL_set_fd(ssl, client_fd) != 1) {
+		LM_ERR("init_ssl_conn(): Failed to set up TLS connection.");
```

Comment:
Can you divulge any more about why it failed? Does the error code hint a reason?